### PR TITLE
Add randomization for molecular splitters

### PR DESCRIPTION
**Pull request by SteshinSS**
# Randomization for molecular splitters

## Description

In my recent work, I wanted to make cross-validation with scaffold splitting. This would allow me to test my model on different splits quite different from each other. However, I found that the `seed` parameter in `ButinaSplitter` and `ScaffoldSplitter` doesn't work. Here is the fix:

- Random seed now works for `ButinaSplitter` and `ScaffoldSplitter`
- `MolecularWeightSplitter` has a `seed` parameter as well, but it is not clear from docs that it's not working. I changed the parameter's description.
- There was a little bug in `FingerprintSplitter`, so even if you set `frac_test` to be 0.0, it would sometimes put one molecule in a test. I fixed it.
- Such molecular splitters have no guarantees on result dataset sizes, so they may vary. It become obvious with the random seed parameter, so I changed several tests that assumed known sizes.
- I remove secret `TimeSplitterPDBbind`. It doesn't exist in docs, it goes under a commentary "Not well supported splitters" and nobody has edited the code for the last 4-5 years, so it seems to be a dead code.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) (`TimeSplitterPDBbind`)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rbharath** [comment]
I've just kicked off a unit test run. Will take a closer look at this PR once the run completes!

**Comment by rbharath** [comment]
Looking at the CI, it looks like we have some additional failures (we should have about 24 tests passing). Could you go through the failures and see if some of them are related to the splitter changes? If you can make it to the developer calls, I'd be happy to help you do this!

**Comment by SteshinSS** [comment]
I compared failed tests to similar [PR](https://github.com/deepchem/deepchem/pull/3038). I changed `ScaffoldSplitter`, so it changed several tests that train models on different splits. Some of them were flaky enough, so they failed. I changed their assertion, so it should be better now. Let's see. 

I looked through each failed test, and It seems that the broken CI is not that broken at all! Windows tests should be either fixed or turned off, and there are a lot of meaningless warnings cluttering the log, but if fixed, the flaky tests shouldn't be so hard to correct. I think people are intimidated by huge warnings, but it's a job for another PR :)

File: deepchem/splits/splitters.py [hunk 1]
```diff
@@ -1408,28 +1456,25 @@ def split(
...
+    train_inds, valid_inds, test_inds = _select_clusters(
+        scaffold_sets, expected_train_size, expected_valid_size,
+        expected_test_size)
+
     return train_inds, valid_inds, test_inds
 
   def generate_scaffolds(self,
                          dataset: Dataset,
-                         log_every_n: int = 1000) -> List[List[int]]:
+                         log_every_n: Optional[int] = 1000) -> List[List[int]]:
```
Comment (rbharath) [inline deepchem/splits/splitters.py:1477]: Why is this made optional? Does changing the type break backwards compatibility?

**Review by rbharath — COMMENTED** [review]
Thanks for the PR! As a general comment, there are a lot of changes in this PR and it's hard for me to judge correctness of the PR as a result. 

`ScaffoldSplitter` is a core class and I'm not sure if this modification would change its behavior in an unexpected way. We should also not delete a class in this PR. I'd encourage you to consider simplifying this PR to focus on one change (`ButinaSplitter` or `ScaffoldSplitter` for example)

I'd encourage you to come by one of the DeepChem developer calls to discuss! (https://forum.deepchem.io/t/joining-the-deepchem-developer-calls/354).

File: deepchem/splits/splitters.py [hunk 2]
```diff
@@ -1408,28 +1456,25 @@ def split(
       Each indices is a list of integers.
     """
     np.testing.assert_almost_equal(frac_train + frac_valid + frac_test, 1.)
-    scaffold_sets = self.generate_scaffolds(dataset)
```
[the remark this answers is not in the document]
Comment (SteshinSS) [reply deepchem/splits/splitters.py:1411]: Yes, it would. It's random already, but I'm adding a possibility to manage that randomness.
Because of it, each test that uses `ScaffoldSplitter` will have slightly different splits, so the result metrics could be different. That is why some tests are failed.

File: deepchem/splits/splitters.py (hunk 1 above)
Comment (SteshinSS) [reply deepchem/splits/splitters.py:1477]: Look at the [call of this function](https://github.com/deepchem/deepchem/blob/master/deepchem/splits/splitters.py#L1411). There was a minor bug, so `log_every_n` wasn't used. I fixed that so it needed to be made optional.

File: deepchem/splits/splitters.py [hunk 3]
```diff
@@ -1360,8 +1410,6 @@ class ScaffoldSplitter(Splitter):
   ... dataset = dc.data.DiskDataset.from_numpy(X=Xs,y=Ys,w=np.zeros(len(data_test)),ids=data_test)
   >>> scaffoldsplitter = dc.splits.ScaffoldSplitter()
   >>> train,test = scaffoldsplitter.train_test_split(dataset)
-  >>> train
```
[the remark this answers is not in the document]
Comment (SteshinSS) [reply deepchem/splits/splitters.py:1363]: `ScaffoldSplitter` behavior is random and depends the  on `seed` parameter. We can't predict the exact content of the `train`, so doctest will fail if we set any outcome in the example. We can fix `seed` in this example, but I am afraid it will be broken in the feature when we update numpy version for instance.

File: docs/source/get_started/examples.rst [hunk 4]
```diff
@@ -73,7 +73,7 @@ First, we'll load the dataset with :func:`load_delaney() <deepchem.molnet.load_d
     >>> assert train_scores['mean-pearson_r2_score'] > 0.7, train_scores
     >>>
     >>> valid_scores = model.evaluate(valid_dataset, [avg_pearson_r2], transformers)
-    >>> assert valid_scores['mean-pearson_r2_score'] > 0.3, valid_scores
+    >>> assert valid_scores['mean-pearson_r2_score'] > 0.2, valid_scores
```
[the remark this answers is not in the document]
Comment (SteshinSS) [reply docs/source/get_started/examples.rst:76]: That is one of the most controversial changes. Some metrics got down and tests stopped working. I tried several hyperparameters, but it didn't help, so I just lower the thresholds.

**Comment by SteshinSS** [comment]
I will leave changes only for `ButinaSplitter` and get back. It should make PR much more simple.

File: deepchem/splits/splitters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/splits/splitters.py:1411]: I believe `ScaffoldSplitter` isn't random at present. The sort is deterministic (https://github.com/deepchem/deepchem/blob/master/deepchem/splits/splitters.py#L1461) and based on size of scaffold splits. 

A change to the behavior of scaffold splitter that introduces new randomness would likely break existing pipelines that expect scaffold split to maintain stability across runs. It might be better as a change to explicitly document and test that `ScaffoldSplitter` is deterministic.

File: deepchem/splits/splitters.py (hunk 1 above)
Comment (rbharath) [reply deepchem/splits/splitters.py:1477]: I think if a default is set it doesn't need to be optional. Making it `Optional` allows for `log_every_n` to be `None`. That isn't something we support at present as a convention. If we make this change, we'd have to make it across the codebase to retain the calling convention

File: deepchem/splits/splitters.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/splits/splitters.py:1363]: See comment above. `ScaffoldSplitter` should be deterministic and tested to be deterministic. That might be a nice fix for a future PR!

File: docs/source/get_started/examples.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply docs/source/get_started/examples.rst:76]: We can't accept a performance regression here. My guess is that the introduced randomness changed the behavior of the test.

**Comment by rbharath** [comment]
Sounds good! It's a good idea for a first PR to stay small to get familiar with DeepChem coding conventions (in general, breaking backwards compatibility or worsening performance is a big deal). You may also find our guidelines for reviewers useful since it highlights what we check for when reviewing https://forum.deepchem.io/t/guidelines-for-deepchem-reviewers/779

**Comment by SteshinSS** [comment]
@rbharath  I assumed Molecular Splitters should allow randomness because [they all have](https://deepchem.readthedocs.io/en/latest/api_reference/splitters.html#molecule-splitters) `seed` parameter, and that is what I personally would like to see, to be able to run cross-validation. What do you think if I add randomness control to `ButinaSplitter`? Would it be useful?

**Comment by peastman** [comment]
The `seed` argument is defined by the parent Splitter class.  It's needed so that all splitters can follow a common API, and other code can invoke them without knowing what particular algorithm to use.  It's up to a particular splitter to decide whether to use it or not.  Some of them are intended to be deterministic.  That isn't a bug.  It just how they work.  The Butina clustering algorithm is deterministic.
